core/src/main/java/org/elasticsearch/index/query/HasChildQueryBuilder.java

```diff
@@ -18,40 +18,104 @@
+import org.elasticsearch.common.xcontent.*;
+import org.elasticsearch.index.fielddata.IndexParentChildFieldData;
+import org.elasticsearch.index.fielddata.plain.ParentChildIndexFieldData;
+import org.elasticsearch.index.mapper.DocumentMapper;
+import org.elasticsearch.index.mapper.internal.ParentFieldMapper;
+import org.elasticsearch.index.query.support.QueryInnerHits;
+import org.elasticsearch.index.search.child.ChildrenConstantScoreQuery;
+import org.elasticsearch.index.search.child.ChildrenQuery;
+import org.elasticsearch.index.search.child.ScoreType;
+import org.elasticsearch.search.fetch.innerhits.InnerHitsContext;
+import org.elasticsearch.search.fetch.innerhits.InnerHitsSubSearchContext;
+import org.elasticsearch.search.internal.SearchContext;
 
 import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
 
+/**
+ * A query builder for <tt>has_child</tt> queries.
+ */
 public class HasChildQueryBuilder extends AbstractQueryBuilder<HasChildQueryBuilder> {
 
+    /**
+     * The queries name
+     */
     public static final String NAME = "has_child";
 
-    private final QueryBuilder queryBuilder;
+    /**
+     * The default cut off point only to evaluate parent documents that contain the matching parent id terms
+     * instead of evaluating all parent docs.
+     */
+    public static final int DEFAULT_SHORT_CIRCUIT_CUTOFF = 8192;
+    /**
+     * The default maximum number of children that are required to match for the parent to be considered a match.
+     */
+    public static final int DEFAULT_MAX_CHILDREN = Integer.MAX_VALUE;
+    /**
+     * The default minimum number of children that are required to match for the parent to be considered a match.
+     */
+    public static final int DEFAULT_MIN_CHILDREN = 0;
+
+    private final QueryBuilder query;
 
-    private String childType;
+    private final String type;
 
-    private String scoreType;
+    private ScoreType scoreType = ScoreType.NONE;
 
-    private Integer minChildren;
+    private int minChildren = DEFAULT_MIN_CHILDREN;
 
-    private Integer maxChildren;
+    private int maxChildren = DEFAULT_MAX_CHILDREN;
 
-    private Integer shortCircuitCutoff;
+    private int shortCircuitCutoff = DEFAULT_SHORT_CIRCUIT_CUTOFF;
 
-    private QueryInnerHitBuilder innerHit = null;
+    private QueryInnerHits queryInnerHits;
 
-    static final HasChildQueryBuilder PROTOTYPE = new HasChildQueryBuilder(null, null);
+    static final HasChildQueryBuilder PROTOTYPE = new HasChildQueryBuilder("", EmptyQueryBuilder.PROTOTYPE);
 
-    public HasChildQueryBuilder(String type, QueryBuilder queryBuilder) {
-        this.childType = type;
-        this.queryBuilder = queryBuilder;
+    public HasChildQueryBuilder(String type, QueryBuilder query, Integer maxChildren, Integer minChildren, Integer shortCircuitCutoff, ScoreType scoreType, QueryInnerHits queryInnerHits) {
+        this(type, query);
+        scoreType(scoreType);
+        this.scoreType = scoreType;
```

Comment:
I don't understand why there is either a setter (with null check & exception) here and then direct assignment to fields. Using either one of these would be fine I think. Same for the other options in this constructor.